pep-0692.rst

```diff
@@ -0,0 +1,344 @@
+
+::
+
+    class Book(TypedDict):
+        genre: str
+        pages: int
+    
+    TypedDictUnion = Movie | Book
+
+    def foo(**kwargs: **TypedDictUnion) -> None: ...  # WRONG! Cannot use a union of TypedDicts to type **kwargs
+
+
+Additional keys
+---------------
+
+A big advantage of using ``**kwargs`` is the ability to pass arbitrary dictionaries containing keyword-value pairs to the function and let the function pick and use only the ones it needs. However, currently, there is no way to define a ``TypedDict`` that would allow additional keys, so supporting this feature would require an addition to the ``TypedDict`` specification. If this limitation blocks the acceptance of this PEP, then the PEP's scope would have to be extended or a new PEP proposing additional keys for ``TypedDict`` would have to be proposed.
+
+Grammar Changes
+===============
+This PEP requires a grammar change so that the double asterisk syntax is allowed for ``**kwargs`` annotations. The proposed change is to extend the ``kwds`` rule as follows:
+
+Before:
+
+::
+
+    kwds: '**' param_no_default 
+
+
+After:
+
+::
+
+    kwds:
+        | '**' param_no_default_double_star_annotation
+        | '**' param_no_default
+
+    param_no_default_double_star_annotation:
+        | param_double_star_annotation & ')'
+
+    param_double_star_annotation: NAME double_star_annotation
+
+    double_star_annotation: ':' double_star_expression
+
+    double_star_expression: '**' expression
+
+A new AST node needs to be created so that type checkers can differentiate the semantics of the new syntax and take into account that the ``**kwargs`` should be unpacked. The proposition is to add a new ``DoubleStarred`` AST node. Then, an AST node for the function defined as:
+
+::
+
+    def foo(**kwargs: **Movie): ...
+
+should look as below:
+
+::
+
+    FunctionDef(
+      name='foo',
+      args=arguments(
+        posonlyargs=[],
+        args=[],
+        kwonlyargs=[],
+        kw_defaults=[],
+        kwarg=arg(
+          arg='kwargs',
+          annotation=DoubleStarred(
+            value=Name(id='Movie', ctx=Load()),
+            ctx=Load())),
+        defaults=[]),
+      body=[
+        Expr(
+          value=Constant(value=Ellipsis))],
+      decorator_list=[])],
+    type_ignores=[])
+
+The runtime annotations should be consistent with the AST. Continuing the previous example:
+
+::
+
+    >>> def foo(**kwargs: **Movie): ...
+    ...
+    >>> foo.__annotations__
+    {'kwargs': **Movie}
+
+In order to make the runtime behaviour consistent with the AST as shown above we propose to:
```

Comment:
We shouldn't specify what goes in typing.py: that's an implementation detail. But we *should* specify what the `**` operator does, which is call the `__unpack__` special method. We should make it explicit that `def f(**kwargs: **T):` is equivalent to `def f(**kwargs: T.__unpack__()):`.